tracker-smoothtrack/ftnoir_tracker_smoothtrack.cpp

```diff
@@ -0,0 +1,184 @@
+}
+
+void smoothtrack::run()
+{
+    // Connect to the iOS device
+    if (!connect_to_device())
+    {
+        return;
+    }
+
+    // Wrap the native socket in a QTcpSocket
+    sock = new QTcpSocket();
+    if (!sock->setSocketDescriptor(usbmuxd_fd))
+    {
+        disconnect_from_device();
+        return;
+    }
+
+    QByteArray buffer;
+    buffer.resize(sizeof(last_recv_pose));
+
+    while (!isInterruptionRequested())
+    {
+        // Wait for data to be available
+        if (!sock->waitForReadyRead(100))
+        {
+            // Check if we're still connected
+            if (sock->state() != QAbstractSocket::ConnectedState)
+            {
+                break;
+            }
+            continue;
+        }
+
+        // Read available data
+        while (sock->bytesAvailable() >= static_cast<qint64>(sizeof(last_recv_pose2)))
+        {
+            QMutexLocker foo(&mutex);
+
+            bool ok = false;
+
+            // Read exactly 48 bytes (6 doubles)
+            const qint64 sz = sock->read(reinterpret_cast<char*>(last_recv_pose2), sizeof(double[6]));
+
+            if (sz == sizeof(double[6]))
+            {
+                ok = true;
+
+                // Validate the data (check for NaN and Infinite values)
+                for (unsigned i = 0; i < 6; i++)
+                {
+                    int val = std::fpclassify(last_recv_pose2[i]);
+                    if (val == FP_NAN || val == FP_INFINITE)
+                    {
+                        ok = false;
+                        break;
+                    }
+                }
+            }
+
+            if (ok)
+            {
+                for (unsigned i = 0; i < 6; i++)
+                    last_recv_pose[i] = last_recv_pose2[i];
+            }
+        }
+    }
+
+    disconnect_from_device();
+}
+
+module_status smoothtrack::start_tracker(QFrame*)
+{
+    // Start the worker thread which will handle the usbmuxd connection
+    start();
+
+    // Give the thread a moment to try connecting
+    msleep(100);
+
+    // Check if connection was successful
+    if (usbmuxd_fd < 0)
+    {
+        return error(tr("Can't connect to iOS device. Make sure:\n"
+                        "1. SmoothTrack app is running on your iOS device\n"
+                        "2. Device is connected via USB\n"
+                        "3. usbmuxd is running\n"
+                        "4. Port number matches SmoothTrack settings"));
+    }
+
+    return status_ok();
+}
+
+void smoothtrack::data(double* data)
+{
+    QMutexLocker foo(&mutex);
+    for (int i = 0; i < 6; i++)
+        data[i] = last_recv_pose[i];
+
+    int values[] = {
+        0, 90, -90, 180, -180,
+    };
+    int indices[] = {
+        s.add_yaw,
+        s.add_pitch,
+        s.add_roll,
+    };
+
+    for (int i = 0; i < 3; i++)
+    {
+        const int k = indices[i];
+        if (k >= 0 && k < std::distance(std::begin(values), std::end(values)))
+            data[Yaw + i] += values[k];
+    }
```

Comment:
With the upcoming addition of a "camera offset", this part will be redundant and should be removed.

Reply:
And the expression should be `std::size(values)`.